crates/compiler/src/llvm/data_layout.rs

```diff
@@ -0,0 +1,1415 @@
+
+use chumsky::{
+    error::Simple,
+    prelude::{choice, just},
+    Parser,
+};
+use ltc_errors::compile::{Error, Result};
+
+use crate::constant::BYTE_SIZE;
+
+/// Information about the expected data-layout for this module.
+///
+/// # Defaulting
+///
+/// LLVM starts with a default specification of the data-layout that is possibly
+/// overridden by the data-layout string. This struct implements this behavior,
+/// so if you want a defaulted layout, either call [`DataLayout::new`] with an
+/// empty string, or use the [`Default`] instance.
+#[derive(Clone, Debug, PartialEq)]
+pub struct DataLayout {
+    /// The endianness used in this data layout.
+    pub endianness: Endianness,
+
+    /// The mangling scheme used by this data layout.
+    pub mangling: Mangling,
+
+    /// The natural alignment of the stack in bits.
+    pub stack_alignment: usize,
+
+    /// The index of the address space that corresponds to program memory.
+    pub program_address_space: usize,
+
+    /// The index of the address space that corresponds to globals.
+    pub global_address_space: usize,
+
+    /// The index of the address space for allocations.
+    pub alloc_address_space: usize,
+
+    /// The layout of pointers.
+    pub pointer_layouts: Vec<PointerLayout>,
+
+    /// The layout of the various integer types.
+    pub integer_layouts: Vec<IntegerLayout>,
+
+    /// The layout of the various vector types.
+    pub vector_layouts: Vec<VectorLayout>,
+
+    /// The layout of the various floating-point types.
+    pub float_layouts: Vec<FloatLayout>,
+
+    /// The layout of aggregate types.
+    pub aggregate_layout: AggregateLayout,
+
+    /// The layout of function pointers.
+    pub function_pointer_layout: FunctionPointerLayout,
+
+    /// The integer widths natively supported by the CPU in this layout.
+    pub native_integer_widths: NativeIntegerWidths,
+
+    /// The address space numbers in which pointers should be treated as
+    /// non-integral.
+    pub nointptr_address_spaces: NonIntegralPointerAddressSpaces,
+}
+
+impl DataLayout {
+    /// Constructs a new data layout description from the provided
+    /// `layout_string`.
+    ///
+    /// If any portion of the data layout specification is left unspecified,
+    /// then the default data layout specification is used in its place as
+    /// described [here](https://llvm.org/docs/LangRef.html#data-layout). In
+    /// addition, we:
+    ///
+    /// - Default to 32 and 64-bit native integer widths.
+    /// - Default to independent function pointers aligned to 64 bits.
+    /// - Default to the ELF mangling scheme if none is specified.
+    ///
+    /// # Errors
+    ///
+    /// - [`Error::InvalidDataLayoutSpecification`] if the provided
+    ///   `layout_string` cannot be parsed as a data layout specification.
+    pub fn new(layout_string: &str) -> Result<Self> {
+        let parts = layout_string.split('-');
+
+        let mut endianness = None;
```

Comment:
Could probably make this a bit cleaner by squishing this all down into a struct instantiation, and then overriding `that_object.property` below. That would also elide the final object construction.